scanner/src/main/kotlin/storages/Sw360Storage.kt

```diff
@@ -0,0 +1,143 @@
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.module.kotlin.readValue
+
+import java.nio.file.Path
+
+import kotlin.io.path.createTempDirectory
+
+import org.eclipse.sw360.antenna.http.HttpClientFactoryImpl
+import org.eclipse.sw360.antenna.http.config.HttpClientConfig
+import org.eclipse.sw360.antenna.sw360.client.adapter.AttachmentUploadRequest
+import org.eclipse.sw360.antenna.sw360.client.adapter.SW360Connection
+import org.eclipse.sw360.antenna.sw360.client.adapter.SW360ConnectionFactory
+import org.eclipse.sw360.antenna.sw360.client.config.SW360ClientConfig
+import org.eclipse.sw360.antenna.sw360.client.rest.resource.attachments.SW360AttachmentType
+import org.eclipse.sw360.antenna.sw360.client.rest.resource.releases.SW360Release
+import org.eclipse.sw360.antenna.sw360.client.utils.SW360ClientException
+
+import org.ossreviewtoolkit.model.Failure
+import org.ossreviewtoolkit.model.Identifier
+import org.ossreviewtoolkit.model.Result
+import org.ossreviewtoolkit.model.ScanResult
+import org.ossreviewtoolkit.model.ScanResultContainer
+import org.ossreviewtoolkit.model.Success
+import org.ossreviewtoolkit.model.config.Sw360StorageConfiguration
+import org.ossreviewtoolkit.model.jsonMapper
+import org.ossreviewtoolkit.model.yamlMapper
+import org.ossreviewtoolkit.scanner.ScanResultsStorage
+import org.ossreviewtoolkit.utils.collectMessagesAsString
+import org.ossreviewtoolkit.utils.log
+
+/**
+ * The SW360 storage back-end uses the sw360-client library in order to read/add attachments from the configured
+ * SW360 instance.
+ */
+class Sw360Storage(
+    sw360Configuration: Sw360StorageConfiguration
+) : ScanResultsStorage() {
+    private val sw360JsonMapper = jsonMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+    private val sw360ConnectionFactory = createSw360Connection(sw360Configuration, sw360JsonMapper)
+    private val releaseClient = sw360ConnectionFactory.releaseAdapter
+
+    override fun readFromStorage(id: Identifier): Result<ScanResultContainer> =
+        try {
+            val scanResults = releaseClient.getSparseReleaseByNameAndVersion(createReleaseName(id), id.version)
+                .flatMap { releaseClient.getReleaseById(it.releaseId) }
+                .map { getScanResultOfRelease(it) }
+                .orElse(emptyList())
+                .map { path ->
+                    yamlMapper.readValue<ScanResult>(path.toFile())
+                }
+            Success(ScanResultContainer(id, scanResults))
+        } catch (e: SW360ClientException) {
+            val message = "Could not read scan results for ${id.toCoordinates()} in SW360."
+            log.info { message }
+            Failure(message)
+        }
+
+    override fun addToStorage(id: Identifier, scanResult: ScanResult): Result<Unit> =
+        try {
+            val cachedScanResult = tmpStorageFile(id)
+            yamlMapper.writeValue(cachedScanResult, scanResult)
```

Comment:
Again, the file and the temporary directory need to be deleted after the upload.